server/sonar-server/src/main/java/org/sonar/server/computation/step/NewCoverageMeasuresStep.java

```diff
@@ -0,0 +1,352 @@
+    for (Counter counter : nonEmptyCounters) {
+      variationsBuilder.setVariation(counter.getPeriodIndex(), measureComputer.compute(counter));
+    }
+    measureRepository.add(counters.getComponent(), metric, newMeasureBuilder().setVariations(variationsBuilder.build()).createNoValue());
+  }
+
+  @Override
+  public String getDescription() {
+    return "Computation of New Coverage measures";
+  }
+
+  /**
+   * Internal class storing the Metric objects for each property of a {@link NewCoverageMetricKeys} instance
+   */
+  private static final class NewCoverageMetrics {
+    private final Metric coverageLineHitsData;
+    private final Metric conditionsByLine;
+    private final Metric coveredConditionsByLine;
+    private final Metric newLinesToCover;
+    private final Metric newUncoveredLines;
+    private final Metric newConditionsToCover;
+    private final Metric newUncoveredConditions;
+
+    public NewCoverageMetrics(MetricRepository metricRepository, NewCoverageMetricKeys keys) {
+      this.coverageLineHitsData = metricRepository.getByKey(keys.coverageLineHitsData());
+      this.conditionsByLine = metricRepository.getByKey(keys.conditionsByLine());
+      this.coveredConditionsByLine = metricRepository.getByKey(keys.coveredConditionsByLine());
+      this.newLinesToCover = metricRepository.getByKey(keys.newLinesToCover());
+      this.newUncoveredLines = metricRepository.getByKey(keys.newUncoveredLines());
+      this.newConditionsToCover = metricRepository.getByKey(keys.newConditionsToCover());
+      this.newUncoveredConditions = metricRepository.getByKey(keys.newUncoveredConditions());
+    }
+
+  }
+
+  /**
+   * Holds the {@link Counter}s (one for each Period in a specific {@link PeriodsHolder}) for a specific Component.
+   */
+  private static final class Counters {
+    private final Component component;
+    private final List<Counter> counters;
+
+    public Counters(Component component, PeriodsHolder periodsHolder) {
+      this.component = component;
+      this.counters = from(periodsHolder.getPeriods()).transform(PeriodToCounter.INSTANCE).toList();
+    }
+
+    public void analyze(@Nullable Date lineDate, int hits, int conditions, int coveredConditions) {
+      if (lineDate == null) {
+        return;
+      }
+      for (Counter counter : getCounters()) {
+        counter.analyze(lineDate, hits, conditions, coveredConditions);
+      }
+    }
+
+    public Component getComponent() {
+      return component;
+    }
+
+    private Iterable<Counter> getCounters() {
+      return this.counters;
+    }
+
+  }
+
+  public static final class Counter {
+    private final int periodIndex;
+    private final Date date;
+    private Integer newLines;
+    private Integer newCoveredLines;
+    private Integer newConditions;
+    private Integer newCoveredConditions;
+
+    private Counter(int periodIndex, @Nullable Date date) {
```

Comment:
Date should not be marked as Nullable as it can never be null. I suppose you took from previous decorator, but even there it was not possible for the date to be null.